### PR TITLE
Fix box finalization issue

### PR DESCRIPTION
**Pull request by we-ai**
This PR, together with frontend updates, is to solve issue [#625](https://github.com/episphere/connect/issues/625). Below are details:
- Made data updates atomic, by using batch
- Added `shipBatchBoxes` to handle batched boxes
- Cleaned up logics for data processing

### CODE REVIEW

File: utils/biospecimen.js [hunk 1]
```diff
@@ -303,36 +303,50 @@ const biospecimenAPIs = async (req, res) => {
...
+        }
+
+        try {
+          const isShipmentSuccessful = await shipBatchBoxes(boxIdAndShipmentDataArray, siteCode);
+          if (!isShipmentSuccessful) {
+            return res.status(500).json({ message: 'Failed to save box data', code: 500 });
+          }
+          return res.status(200).json({ message: 'Success!', code: 200 });
+        } catch (error) {
+          return res.status(500).json({ message: 'Internal Server Error', code: 500 });
```
Comment (anthonypetersen) [inline utils/biospecimen.js:347]: would it be helpful to console.log() the error message for troubleshooting purposes?

File: utils/firestore.js [hunk 2]
```diff
@@ -1096,6 +1096,36 @@ const updateTempCheckDate = async (institute) => {
...
+      .get();
+
+    if (snapshot.size !== 1) {
+      return false;
+    }
+
+    batch.update(snapshot.docs[0].ref, shipmentData);
+  }
+
+  await batch.commit().catch((err) => {
```
Comment (anthonypetersen) [inline utils/firestore.js:1121]: very neat use of batching

File: utils/biospecimen.js [hunk 3]
```diff
@@ -303,36 +303,50 @@ const biospecimenAPIs = async (req, res) => {
...
-            }
-            console.log('box: ' + box)
-            console.log(JSON.stringify(requestData))
-            
-            const exists = await shipBox(box, siteCode, shippingData, trackingNumbers)
-            if(exists === false){
-                return res.status(500).json({message: 'Box does not exist', code:500})
-            }
+        const boxIdToTrackingNumberMap = req.body.boxIdToTrackingNumberMap;
+        const boxIdArray = Object.keys(boxIdToTrackingNumberMap);
```
[the remark this answers is not in the document]
Comment (we-ai) [reply utils/biospecimen.js:308]: Good question. Yes. I'll update to fix exceptions. This question applies to all our APIs: we need to check data before processing.

File: utils/biospecimen.js [hunk 4]
```diff
@@ -303,36 +303,50 @@ const biospecimenAPIs = async (req, res) => {
...
+
+        if (boxIdArray.length === 0 ) {
+            return res.status(400).json(getResponseJSON('Request body is empty!', 400));
         }
-        return res.status(200).json({message: 'Success!', code:200})
+        
+        const { shipBatchBoxes} = require('./firestore');
+        
+        let {boxWithTempMonitor, ...sharedShipmentData} = req.body.shippingData;
```
[the remark this answers is not in the document]
Comment (we-ai) [reply utils/biospecimen.js:316]: Yes. It's necessary.

**Review by Gbarra9 — APPROVED** [review]

**Pull request merged**